RESOURCE_SHARING_AND_ACCESS_CONTROL.md

```diff
@@ -0,0 +1,673 @@
+    org.opensearch.sample.SampleResourceSharingExtension
+    ```
+
+
+## **3. Resource Sharing API Design**
+
+### **Resource Sharing Index **
+
+Each plugin receives its own sharing index, centrally managed by security plugin, which stores **resource access metadata**, mapping **resources to their access control policies**.
+
+
+|**Field**  |**Type** | Description                                                                            |
+|---  |---  |----------------------------------------------------------------------------------------|
+|`source_idx` |String | The system index where the resource is stored.                                         |
+|`resource_id`  |String | Unique ID of the resource within `source_idx`.                                         |
+|`created_by` |Object | Information about the user or backend role that created the resource.                  |
+|`share_with` |Object | Contains multiple objects with **access-levels** as keys and access details as values. |
+
+NOTE: **action-groups** and **access-levels** are used inter-changeably throughout this document, as they carry the same meaning in the context of this feature.
+
+#### **`created_by`**
+
+### This object contains details about the **creator** of the resource.
+
+|**Field**  |**Type** |Description  |
+|---  |---  |---  |
+| user|String |The username of the creator. |
+
+**Example:**
+
+```
+"created_by": {
+   "user": "darshit"
+}
+```
+
+#### **`share_with`**
+
+The `share_with` field **contains multiple objects**, where each **key is an action-group (e.g., `read`, `read_write`)**, and the **value is an object defining access control**.
+
+|Scope Key  |**Type** |Description  |
+|---  |---  |---  |
+|<action-group> |Object |Define access level for its corresponding users, roles, and backend roles  |
+
+**Example:**
+
+```
+"share_with": {
+   "action-group1": {
+      "users": ["user1", "user2"],
+      "roles": ["viewer_role"],
+      "backend_roles": ["data_analyst"]
+   },
+   "action-group2": {
+      "users": ["admin_user"],
+      "roles": ["editor_role"],
+      "backend_roles": ["content_manager"]
+   }
+}
+```
+
+**NOTE**: The action-groups here are synonymous with the existing action-groups in security plugin and can be used to share/revoke resource access.
+
+#### **Each Action-Group in `share_with`**
+
+Each **action-group** entry contains the following access definitions:
+
+|**Field**  |**Type** |Description  |
+|---  |---  |---  |
+|`users`  |Array  |List of usernames granted access under this action-group. |
+|`roles`  |Array  |List of OpenSearch roles granted access under this action-group.  |
+|`backend_roles`  |Array  |List of backend roles granted access under this action-group. |
+
+**Example:**
+
+```
+"action-group1": {
+   "users": ["user1", "user2"],
+   "roles": ["viewer_role"],
+   "backend_roles": ["data_analyst"]
+}
+```
+
+
+### **Example Resource-Sharing Document**
+
+```
+{
+   "source_idx": ".plugins-ml-model-group",
```

Comment:
Isn't this a little outdated?

The index would be called `.plugins-ml-model-group-sharing` and the ids of the docs in this index would match the ids of the docs in `.plugins-ml-model-group` index.

We should audit the use of `source_idx` in javadoc and make sure its accurately used.